app/js/endTest.ts

```diff
@@ -0,0 +1,18 @@
+var pageLeft = false
+
+function pageLeaveAlert() {
+    if (!pageLeft) {
+        alert ('Beware, if you leave the page the test will end!')
+    }
+}
+
+function cancelTest() {
+    if (document.visibilityState === "hidden") {
+        pageLeft = true
+        finishTest()
+    }
+}
+
+document.addEventListener("mouseleave", pageLeaveAlert);
+document.addEventListener("visibilitychange", cancelTest);
+
```

Comment:
A few problems here:

1. Line 1 - declaring a variable with global scope and then using it in the local scope of the functions. This isn't really best practice in general, but especially in a project like this with lots of different TS files that get transpiled into JS and then run on the page, do you know when this line of code will actually run? Will `pageLeft` be initialised at the time when the functions are called? It is hard to know, which is why this is not really trustworthy.

Normally it is always best to pass variables as parameters rather than inheriting them from the global scope.

2. Do you actually need that variable? What does it do?

I welcome your explanations below 👇

Reply:
I've updated this slightly - hopefully works better now?